TCP endpoint compare refactoring

jukkar opened the pull request:
Instead of allocating an endpoint union and then freeing it almost immediately, allocate the ep from stack instead of heap.
This contains also Ravi's SRC/DST rename patch from the TCP2 unit test PR #24346

jukkar committed:
net: tcp2: Avoid extra memory allocation in endpoint cmp

In tcp_endpoint_cmd() we allocate and then almost immediately
destroy the allocated endpoint. This is quite inefficient so
use a endpoint from stack in the compare.

Separate endpoint allocation from value setting so the caller
can decide how the endpoint union values are set.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

ozhuraki reviewed (commented):
@jukkar

I would suggest a following approach here, i.e.:

1.	Externalize the memory management from tcp_endpoint_new()
2.	tcp_endpoint_new() -> tcp_endpoint_set()

What do you think?

jukkar commented:
> What do you think?

Agreed, the allocation and setting are not needed in the same function. I can look into it.

jukkar committed:
tests: net: tcp2: Add more debugging info

It is useful to know which test fails to semaphore timeout so
add line number of the failing test to assert print.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Separated the allocation and setup of endpoint union.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -58,38 +58,57 @@ static size_t tcp_endpoint_len(sa_family_t af)
+}
 
-		ep->sin.sin_port = src ? th->th_sport : th->th_dport;
+static union tcp_endpoint *tcp_endpoint_set(union tcp_endpoint *ep,
```
I would suggest to change the return type to bool. This way it would be easier to do the error handling more clear and intuitive inside this function and in places of its usage.

jukkar replied:
I did not want to change the current behaviour too much. So new `tcp_endpoint_set()` can be put to same place where `tcp_endpoint_new()` was earlier. Actually we do not seem to handle the out-of-mem case very well in current code either. Hmm, I will send another PR that adds more error checks, will check also would it be better to change the return type.

ozhuraki replied:
Sure, this can be done later. These (handling out or memory and more explicit/clear error handling) were my points.

Alternative approach to consider later is just to embed endpoints into struct tcp and to drop the heap allocations.

jukkar replied:
> Alternative approach to consider later is just to embed endpoints into struct tcp and to drop the heap allocations.

Yes, that would be better way indeed.

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -689,15 +708,15 @@ int net_tcp_get(struct net_context *context)
-	union tcp_endpoint *ep_new = tcp_endpoint_new(pkt, which);
-	bool is_equal = memcmp(ep, ep_new, tcp_endpoint_len(ep->sa.sa_family)) ?
-		false : true;
+	union tcp_endpoint ep_tmp = { 0 }, *ep_ptr;
```
Perhaps, ep_ptr is extra here, &ep_tmp can be passed right to memcmp().

ozhuraki commented on subsys/net/ip/tcp2.c:
```diff
@@ -789,8 +808,8 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
 
-	conn->dst = tcp_endpoint_new(pkt, TCP_EP_SRC);
-	conn->src = tcp_endpoint_new(pkt, TCP_EP_DST);
+	conn->dst = tcp_endpoint_set(tcp_endpoint_new(pkt), pkt, TCP_EP_SRC);
```
Per your judgement, consider the first comment (tcp_enpoint_set() -> bool), might make the error/reporting handling more clear.

ozhuraki reviewed (approved):
LGTM.

A few improvements per your judgement.
